### PR TITLE
exceptions: specialize base exceptions (SC-1436)

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed commit msg:

```
exceptions: specialize exceptions

Add custom exceptions to improve consumer code handling abilities.

Fixes: #263
```

## Additional context:

Addresses #263

### CODE REVIEW

**Review by TheRealFalcon — COMMENTED** [review]
Do you think we can get rid of `PycloudlibException`? I don't see why we need both that and `PycloudlibError`. Do any of these errors _need_ to inherit from RuntimeError? Looking at the error file, it's not immediately obvious which one I would choose if I needed to throw a new pycloudlib related error. 

Additionally, I think it might be worth having a `PycloudlibTimeoutError` for operations where we expect to wait for something to happen but it never gets there.

Otherwise, PR looks good!

File: pycloudlib/azure/cloud.py [hunk 1]
```diff
@@ -873,21 +874,19 @@ def get_instance(self, instance_id, search_all=False):
...
-                raise Exception(
-                    "The image {} was already deleted".format(instance_id)
+                raise PycloudlibError(
+                    f"The image {instance_id} was already deleted"
                 )
 
             return instance
 
-        raise Exception("Could not find {}".format(instance_id))
+        raise InstanceNotFoundError("Could not find {}".format(instance_id))
```
Comment (aciba90) [inline pycloudlib/azure/cloud.py:889]:
```suggestion
        raise InstanceNotFoundError(resource_id=instance_id)
```

**Comment by aciba90** [comment]
Thanks, @TheRealFalcon.

> Do you think we can get rid of `PycloudlibException`? I don't see why we need both that and `PycloudlibError`. Do any of these errors _need_ to inherit from RuntimeError? Looking at the error file, it's not immediately obvious which one I would choose if I needed to throw a new pycloudlib related error.

`PycloudlibException` mimics `Exception`. All custom exceptions must inherit from it, allowing consumer code to catch any exception. Pycloudlib should not raise it.

`PycloudlibError` mimics `RuntimeError`.

Having both exceptions allow clients to catch any exception or any uncategorized exception, offering more rich semantics.

I have reflected that in the documentation.

> Additionally, I think it might be worth having a `PycloudlibTimeoutError` for operations where we expect to wait for something to happen but it never gets there.

I have added that one and went forward, specializing the rest of `RuntimeError` exceptions and improved the documentation related to exceptions. I believe this solves #282.

**Comment by TheRealFalcon** [comment]
Sorry, one more:
https://github.com/canonical/pycloudlib/blob/main/pycloudlib/instance.py#L481

I think that needs to be changed to a `PycloudlibTimeoutError`. 

#263 also has an issue with https://github.com/canonical/pycloudlib/blob/main/pycloudlib/instance.py/#L64 , but I don't think it makes sense to change that one. If we get to line 64, the execute command has worked, so we don't have a remote connection issue. Rather, the `cat` command itself has failed meaning `boot_id` does not exist or we had some trouble reading it. Given

> This exception is raised when a system function returns a system-related error, including I/O failures such as “file not found” or “disk full” (not for illegal argument types or other incidental errors).

I think `OSError` is reasonable in that case.

**Comment by aciba90** [comment]
Agreed, @TheRealFalcon.

(I missed this) About `PycloudlibError` inheriting from `RuntimeError` in:

https://github.com/canonical/pycloudlib/blob/393da249ce7ee4479a74cfd1183236f573b13025/pycloudlib/errors.py#L19

The reason was to keep backwards compatibility, because I wanted to split the changes in two PRs because I did not have the versioning / publishing tooling.

Now that we have it and we are going to do the change at once, we could drop that inheritance.

Do you think it is worth removing it?

**Review by TheRealFalcon — APPROVED** [review]
> Do you think it is worth removing it?

Yes, since we're bumping major version, I don't think it's too much to ask people to update their exceptions appropriately.

+1 from me either way though. Feel free to make the change (or not) and merge.

**Comment by aciba90** [comment]
> > Do you think it is worth removing it?
> 
> Yes, since we're bumping major version, I don't think it's too much to ask people to update their exceptions appropriately.
> 
> +1 from me either way though. Feel free to make the change (or not) and merge.

Thanks, I will do the change.

**Pull request merged**